UniversalAlgebra/Structures/Entailment.lagda

```diff
@@ -0,0 +1,63 @@
+---
+layout: default
+title : Structures.Terms
+date : 2021-07-02
+author: [the agda-algebras development team][]
+---
+
+
+\begin{code}
+
+{-# OPTIONS --without-K --exact-split --safe #-}
+
+module Structures.Entailment where
+
+open import Agda.Primitive   using    ( _⊔_   ;  Level
+                                      ; lsuc           )
+                             renaming ( Set   to Type
+                                      ; lzero to ℓ₀    )
+open import Relation.Unary   using    ( Pred  ;  _∈_   )
+open import Data.Nat         using    ( ℕ              )
+open import Data.Fin.Base    using    ( Fin            )
+open import Data.Product     using    ( _,_   ;  _×_   )
+                             renaming ( proj₁    to  fst
+                                      ; proj₂    to  snd   )
+
+-- -- Imports from agda-algebras --------------------------------------
+open import Overture.Preliminaries     using ( _≈_ )
+open import Structures.AsRecordsBasic  using ( signature ; structure )
+open import Structures.Terms
+
+open signature
+open structure
+
+ℓ₁ : Level
+ℓ₁ = lsuc ℓ₀
+
+module _ {𝐹 𝑅 : signature}{χ : Level}{X : Type χ} where
+
+ _⊧_≈_ : structure 𝐹 𝑅 → Term X → Term X → Type _
+ 𝑨 ⊧ p ≈ q = 𝑨 ⟦ p ⟧ ≈ 𝑨 ⟦ q ⟧
+
+ _⊧_≋_ : Pred(structure 𝐹 𝑅) ℓ₀ → Term X → Term X → Type _
+ 𝒦 ⊧ p ≋ q = {𝑨 : structure 𝐹 𝑅} → 𝒦 𝑨 → 𝑨 ⊧ p ≈ q
+
+ -- Theories
+ Th : Pred (structure 𝐹 𝑅) ℓ₀ → Pred(Term X × Term X) (ℓ₁ ⊔ χ)
```

Comment:
Why uncurried? i.e. why do you use a unary predicate on the right instead of a binary relation? The latter would be more Agda-friendly and require less packing/unpacking in various definitions.

Reply:
The other way to go, is to consider an 'equation' as a real object worthy of a name, and make it into either a record or a 1-constructor data. Both would help with type inference and understandability of the goals.